### PR TITLE
added component resolver

### PR DESCRIPTION
**Pull request by Polleke007**

### CODE REVIEW

File: src/components/ctf-components/ctf-text-banner/__generated/ctf-text-banner.generated.ts [hunk 1]
```diff
@@ -0,0 +1,45 @@
+import * as Types from '../../../../lib/__generated/graphql.types';
+
+import { useQuery, UseQueryOptions } from '@tanstack/react-query';
+import { customFetcher } from '@/lib/fetchConfig';
+export type TextBannerFieldsFragment = { __typename: 'TextBanner', title?: string | null, description?: string | null, sys: { __typename?: 'Sys', id: string } };
```
Comment (DavithkbY) [inline src/components/ctf-components/ctf-text-banner/__generated/ctf-text-banner.generated.ts:5]: If you have ? you don't need to specify null anymore I thinkl

File: src/components/utils/component-resolver.tsx [hunk 2]
```diff
@@ -1,27 +1,97 @@
...
+}
+
+export const ComponentResolver = (props: Props) => {
+    const { componentProps, inline = false } = props;
+    const { previewActive } = useContentfulContext();
+
+    const { locale } = useContentfulContext();
+
+    // @ts-ignore
+    const ComponentGql = componentGqlMap[componentProps.__typename];
```
Comment (DavithkbY) [inline src/components/utils/component-resolver.tsx:36]: why is this ts ignore here?

File: src/components/utils/component-resolver.tsx [hunk 3]
```diff
@@ -1,27 +1,97 @@
...
+            // We expect exactly these keys to be present in the returned props if the
+            // fragment was not specified for this component
+            return false;
+        }
+
+        return true;
+    }, [ComponentGql, componentProps, props.forceGql]);
+
+    // @ts-ignore
+    const Component = !shouldForceGql && componentMap[componentProps.__typename];
```
Comment (DavithkbY) [inline src/components/utils/component-resolver.tsx:63]: Can we get rid of this ts ignore?

File: src/components/utils/component-resolver.tsx [hunk 4]
```diff
@@ -1,27 +1,97 @@
...
+                            id={componentProps.sys.id}
+                            __typename={componentProps.__typename}
+                            className={props.className}
+                            preview={previewActive}
+                            locale={locale}
+                            previousComponent={previousComponentProp}
+                        />
+                    )}
+                </div>
+        // <CtfTextBannerGql id={componentProps.sys.id} locale={locale} preview={previewActive}></CtfTextBannerGql>
```
Comment (DavithkbY) [inline src/components/utils/component-resolver.tsx:95]: Unneeded commented code?

**Review by DavithkbY — CHANGES_REQUESTED** [review]
Please review the comments

File: src/components/ctf-components/ctf-text-banner/__generated/ctf-text-banner.generated.ts (hunk 1 above)
Comment (ceneva) [reply src/components/ctf-components/ctf-text-banner/__generated/ctf-text-banner.generated.ts:5]: The `?` in typings means the property itself is optional, and only applies to undefined (property doesn't exist), not to null (since null is a value). So the typing is indeed correct if null is an allowed value.
However, when used in optional chaining, this does cover null values as well, maybe that's why this may seem confusing?

File: src/components/utils/component-resolver.tsx (hunk 2 above)
Comment (ceneva) [reply src/components/utils/component-resolver.tsx:36]: removing the @ts-ignore results in errors on the typings.
const ComponentGql = (componentGqlMap as {type})[componentProps.__typename];
Replace {type} with the proper typing on componentGqlMap. I've seen this module is lazy loaded, so best to separate the typings so you can still use them elsewhere.

File: src/components/utils/component-resolver.tsx [hunk 5]
```diff
@@ -1,27 +1,97 @@
...
 
-export const ComponentResolver = () => {
-  return (
-    <div>
-      ComponentResolver TODO: see
-      https://github.com/contentful/template-marketing-webapp-nextjs
-    </div>
-  );
+
+interface index {
```
Comment (ceneva) [inline src/components/utils/component-resolver.tsx:25]: This interface isn't used, nor exported?

File: src/components/ctf-components/ctf-page/ctf-page.tsx [hunk 6]
```diff
@@ -27,64 +32,65 @@ export default async function CtfPage() {
     metaTags.no_follow === true ? 'nofollow' : undefined,
   ].filter((x): x is string => x !== undefined);
 
+  const contentfulData = {
+     sys: { id: 'VL0GACJpBfmSRvACLvMdP'},
```
Comment (koenkicken) [inline src/components/ctf-components/ctf-page/ctf-page.tsx:36]: @Polleke007  what is this? why is this needed?

File: src/components/ctf-components/ctf-text-banner/ctf-text-banner-ggl.tsx [hunk 7]
```diff
@@ -0,0 +1,17 @@
+'use client';
```
Comment (koenkicken) [inline src/components/ctf-components/ctf-text-banner/ctf-text-banner-ggl.tsx:1]: what is this `-ggl` suffix?

File: src/components/utils/component-resolver.tsx (hunk 2 above)
Comment (Polleke007) [reply src/components/utils/component-resolver.tsx:36]: No idea why but for some reason the error is gone

File: .env [hunk 8]
```diff
@@ -0,0 +1,5 @@
+CONTENTFUL_SPACE_ID=2p89d5j9ulp9
```
[the remark this answers is not in the document]
Comment (Polleke007) [reply .env:1]: yeah it was indeed a accident, should be on the .gitignote on master

File: src/components/ctf-components/ctf-page/ctf-page.tsx (hunk 6 above)
Comment (Polleke007) [reply src/components/ctf-components/ctf-page/ctf-page.tsx:36]: The id of the content that should be retrieved

File: src/components/ctf-components/ctf-text-banner/ctf-text-banner-ggl.tsx (hunk 7 above)
Comment (Polleke007) [reply src/components/ctf-components/ctf-text-banner/ctf-text-banner-ggl.tsx:1]: I think it's used internal for retrieving the initial data before it's cached (not sure though), it's also used in the contentful examples so I sticked with it

File: src/components/ctf-components/ctf-page/ctf-page.tsx (hunk 6 above)
Comment (koenkicken) [reply src/components/ctf-components/ctf-page/ctf-page.tsx:36]: @Polleke007  But why is this hardcoded? This should be configured in my 'AmexioPage' (see getAmexioPage method).
It should not be done in this way

File: src/components/ctf-components/ctf-text-banner/ctf-text-banner-ggl.tsx (hunk 7 above)
Comment (koenkicken) [reply src/components/ctf-components/ctf-text-banner/ctf-text-banner-ggl.tsx:1]: You mean `-gql` suffix then?

File: next.config.mjs [hunk 9]
```diff
@@ -1,15 +1,25 @@
...
+        protocol: 'https',
+        hostname: 'images.ctfassets.net',
+        port: '',
+        pathname: '/**',
+      },
+    ],
+  },
+  env: {
+    CONTENTFUL_SPACE_ID: '2p89d5j9ulp9',
+    CONTENTFUL_PREVIEW_ACCESS_TOKEN: 'mTHrlfx3vWf2wCr7VYS0c7WsfeD41KxVMyqXAfs7YXE',
```
[the remark this answers is not in the document]
Comment (Polleke007) [reply next.config.mjs:15]: how do you know they're exposed and not only set in the server side?

[the remark this answers is not in the document]
Comment (koenkicken) [reply next.config.mjs:15]: These should not be committed to git 😉 this should be kept secret

File: next.config.mjs
Comment (DavithkbY) [inline next.config.mjs:1]: Can you please create new tokens because by now bots have crawled them @Polleke007
